File: src/tools/resources.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import pino from 'pino';
import * as k8s from '@kubernetes/client-node';
import YAML from 'yaml';
import { getClients } from '../k8s.js';
import { ApplyResourceParams, DeleteResourceParams, GetResourceParams } from '../schemas.js';
import { enforceMutatingGuards, rateLimit } from '../authz.js';

async function listOrGetGeneric(
  clients: ReturnType<typeof getClients>,
  args: { group?: string; version: string; kind: string; name?: string; namespace?: string; labelSelector?: string; fieldSelector?: string; limit?: number }
) {
  const koa = k8s.KubernetesObjectApi.makeApiClient(clients.kubeConfig);
  const apiVersion = args.group ? `${args.group}/${args.version}` : args.version;
  if (args.name) {
    const item = await koa.read({ apiVersion, kind: args.kind, metadata: { name: args.name, namespace: args.namespace || clients.defaultNamespace } });
    return { item } as any;
  }
  const list = await koa.list(apiVersion, args.kind, args.namespace || clients.defaultNamespace, undefined, undefined, undefined, args.fieldSelector, args.labelSelector, args.limit);
  const summary = (list.items as any[]).map((it: any) => ({ apiVersion: it.apiVersion, kind: it.kind, name: it.metadata?.name, namespace: it.metadata?.namespace, uid: it.metadata?.uid, creationTimestamp: it.metadata?.creationTimestamp }));
  return { items: summary };
}

export function registerResourceTools(server: McpServer, logger: pino.Logger) {
  const clients = getClients(logger);

  server.tool(
    'resources.get',
    'Get or list arbitrary resources by GVK',
    GetResourceParams.shape,
    async (args) => {
      rateLimit('resources.get');
      const result = await listOrGetGeneric(clients, args);
      return { content: [{ type: 'text', text: JSON.stringify(result) }] };
    }
  );

  server.tool(
    'resources.apply',
    'Apply manifest YAML (server-side apply by default)',
    ApplyResourceParams.shape,
    async (args) => {
      rateLimit('resources.apply');
      const docs = YAML.parseAllDocuments(args.manifestYAML).map((d) => d.toJSON());
      const koa = k8s.KubernetesObjectApi.makeApiClient(clients.kubeConfig);

      const results: any[] = [];
      for (const obj of docs) {
        if (!obj || typeof obj !== 'object') continue;
        const { apiVersion, kind, metadata } = obj as any;
        if (!apiVersion || !kind || !metadata?.name) {
          results.push({ error: 'Invalid manifest: missing apiVersion/kind/metadata.name' });
          continue;
        }
        enforceMutatingGuards(logger, 'resources.apply', { namespace: metadata.namespace, kind });
        const applied = await k8s.KubernetesObjectApi.makeApiClient(clients.kubeConfig).patch(obj as any, undefined, args.dryRun !== false ? 'All' : undefined, args.fieldManager, true, k8s.PatchStrategy.ServerSideApply);
        results.push({ kind: (applied as any).kind, name: (applied as any).metadata?.name, namespace: (applied as any).metadata?.namespace });
      }
      return { content: [{ type: 'text', text: JSON.stringify({ results }) }] };
    }
  );

  server.tool(
    'resources.delete',
    'Delete a resource by GVK/name',
    DeleteResourceParams.shape,
    async (args) => {
      rateLimit('resources.delete');
      enforceMutatingGuards(logger, 'resources.delete', { namespace: args.namespace, kind: args.kind });
      const koa = k8s.KubernetesObjectApi.makeApiClient(clients.kubeConfig);
      const apiVersion = args.group ? `${args.group}/${args.version}` : args.version;
      const status = await koa.delete({ apiVersion, kind: args.kind, metadata: { name: args.name, namespace: args.namespace || clients.defaultNamespace } }, undefined, args.dryRun !== false ? 'All' : undefined, args.gracePeriodSeconds, undefined, args.propagationPolicy);
      return { content: [{ type: 'text', text: JSON.stringify({ status: (status as any).status || 'Success' }) }] };
    }
  );
}

File: internal/tools/resources.go
package tools

import (
	"context"
	"encoding/json"
	"errors"
	"strings"

	"github.com/example/mcp-k8s-server-go/internal/authz"
	"github.com/example/mcp-k8s-server-go/pkg/k8s"
	"github.com/example/mcp-k8s-server-go/pkg/mcp"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/apimachinery/pkg/types"
)

func RegisterResources(reg *mcp.Registry, k *k8s.Clients) {
	if k == nil {
		notReady := func(ctx context.Context, _ json.RawMessage) (any, error) {
			return nil, errors.New("Kubernetes client not initialized yet")
		}
		reg.Register(mcp.Tool{Name: "resources-get", Description: "Get or list arbitrary resources by GVK", DirectResult: true, Handler: notReady})
		reg.Register(mcp.Tool{Name: "resources-apply", Description: "Apply manifest YAML (server-side apply by default)", DirectResult: true, Handler: notReady})
		reg.Register(mcp.Tool{Name: "resources-delete", Description: "Delete a resource by GVK/name", DirectResult: true, Handler: notReady})
		return
	}
	// resources-get
	reg.Register(mcp.Tool{
		Name:         "resources-get",
		Description:  "Get or list arbitrary resources by GVK",
		DirectResult: true,
		Handler: func(ctx context.Context, params json.RawMessage) (any, error) {
			_ = authz.RateLimit("resources-get", 10, 5)
			var p struct {
				Group         *string
				Version       string
				Kind          string
				Name          *string
				Namespace     *string
				LabelSelector *string
				FieldSelector *string
				Limit         *int
			}
			if err := json.Unmarshal(params, &p); err != nil {
				return nil, err
			}
			gvk := schema.GroupVersionKind{Group: func() string {
				if p.Group == nil {
					return ""
				}
				return *p.Group
			}(), Version: p.Version, Kind: p.Kind}
			gvr, err := k.ResolveResource(gvk)
			if err != nil {
				return nil, err
			}
			ns := ""
			if p.Namespace != nil {
				ns = *p.Namespace
			} else {
				ns = k.DefaultNamespace
			}
			nri := k.Dynamic.Resource(gvr).Namespace(ns)
			if p.Name != nil && *p.Name != "" {
				item, err := nri.Get(ctx, *p.Name, metav1.GetOptions{})
				if err != nil {
					return nil, err
				}
				return map[string]any{"item": item.Object}, nil
			}
			list, err := nri.List(ctx, metav1.ListOptions{LabelSelector: ptrStr(p.LabelSelector), FieldSelector: ptrStr(p.FieldSelector)})
			if err != nil {
				return nil, err
			}
			var summary []map[string]any
			for _, it := range list.Items {
				summary = append(summary, map[string]any{"apiVersion": it.GetAPIVersion(), "kind": it.GetKind(), "name": it.GetName(), "namespace": it.GetNamespace(), "uid": string(it.GetUID()), "creationTimestamp": it.GetCreationTimestamp()})
			}
			if p.Limit != nil && *p.Limit > 0 && len(summary) > *p.Limit {
				summary = summary[:*p.Limit]
			}
			return map[string]any{"items": summary}, nil
		},
	})

	// resources-apply (server-side apply)
	reg.Register(mcp.Tool{
		Name:         "resources-apply",
		Description:  "Apply manifest YAML (server-side apply by default)",
		DirectResult: true,
		Handler: func(ctx context.Context, params json.RawMessage) (any, error) {
			_ = authz.RateLimit("resources-apply", 10, 5)
			var p struct {
				ManifestYAML string  `json:"manifestYAML"`
				FieldManager *string `json:"fieldManager"`
				DryRun       *bool   `json:"dryRun"`
			}
			if err := json.Unmarshal(params, &p); err != nil {
				return nil, err
			}
			docs := splitYAMLDocs(p.ManifestYAML)
			var results []map[string]any
			for _, d := range docs {
				if strings.TrimSpace(d) == "" {
					continue
				}
				obj, err := k8s.DecodeYAMLToUnstructured([]byte(d))
				if err != nil {
					results = append(results, map[string]any{"error": err.Error()})
					continue
				}
				if err := authz.EnforceMutating("resources-apply", obj.GetNamespace(), obj.GetKind()); err != nil {
					results = append(results, map[string]any{"error": err.Error()})
					continue
				}
				gvk := obj.GroupVersionKind()
				gvr, err := k.ResolveResource(gvk)
				if err != nil {
					results = append(results, map[string]any{"error": err.Error()})
					continue
				}
				ns := obj.GetNamespace()
				ri := k.Dynamic.Resource(gvr).Namespace(ns)
				fm := "mcp-k8s-server"
				if p.FieldManager != nil && *p.FieldManager != "" {
					fm = *p.FieldManager
				}
				dr := []string{}
				if p.DryRun == nil || *p.DryRun {
					dr = []string{"All"}
				}
				applied, err := ri.Patch(ctx, obj.GetName(), types.ApplyPatchType, []byte(d), metav1.PatchOptions{FieldManager: fm, Force: ptrBool(true), DryRun: dr})
				if err != nil {
					results = append(results, map[string]any{"error": err.Error()})
					continue
				}
				results = append(results, map[string]any{"kind": applied.GetKind(), "name": applied.GetName(), "namespace": applied.GetNamespace()})
			}
			return map[string]any{"results": results}, nil
		},
	})

	// resources-delete
	reg.Register(mcp.Tool{
		Name:         "resources-delete",
		Description:  "Delete a resource by GVK/name",
		DirectResult: true,
		Handler: func(ctx context.Context, params json.RawMessage) (any, error) {
			_ = authz.RateLimit("resources-delete", 10, 5)
			var p struct {
				Group               *string
				Version, Kind, Name string
				Namespace           *string
				PropagationPolicy   *string
				GracePeriodSeconds  *int64
				DryRun              *bool
			}
			if err := json.Unmarshal(params, &p); err != nil {
				return nil, err
			}
			gvk := schema.GroupVersionKind{Group: ptrStr(p.Group), Version: p.Version, Kind: p.Kind}
			gvr, err := k.ResolveResource(gvk)
			if err != nil {
				return nil, err
			}
			ns := ptrStr(p.Namespace)
			ri := k.Dynamic.Resource(gvr).Namespace(ns)
			if err := authz.EnforceMutating("resources-delete", ns, p.Kind); err != nil {
				return nil, err
			}
			dr := []string{}
			if p.DryRun == nil || *p.DryRun {
				dr = []string{"All"}
			}
			var pp *metav1.DeletionPropagation
			if p.PropagationPolicy != nil {
				s := *p.PropagationPolicy
				switch s {
				case "Foreground":
					v := metav1.DeletePropagationForeground
					pp = &v
				case "Background":
					v := metav1.DeletePropagationBackground
					pp = &v
				case "Orphan":
					v := metav1.DeletePropagationOrphan
					pp = &v
				}
			}
			opts := metav1.DeleteOptions{DryRun: dr, GracePeriodSeconds: p.GracePeriodSeconds, PropagationPolicy: pp}
			if err := ri.Delete(ctx, p.Name, opts); err != nil {
				return nil, err
			}
			return map[string]any{"status": "Success"}, nil
		},
	})
}

func splitYAMLDocs(s string) []string {
	parts := strings.Split(s, "\n---")
	return parts
}

func ptrStr(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}
func ptrBool(v bool) *bool { b := v; return &b }
